File: frontend/types/approval-step.ts
export type ApprovalStatus =
  | "pending"
  | "open"
  | "approved"
  | "changes_requested"
  | "rejected"
  | "skipped";

export interface ApprovalStep {
  id: string;
  target_type: string;
  target_id: string;
  design_request_id: string;
  deliverable_id?: string | null;
  step_order: number;
  step_name: string;
  approver_role?: string | null;
  approver_name?: string | null;
  status: ApprovalStatus;
  decision_reason?: string | null;
  due_at?: string | null;
  decided_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ApprovalStepCreateInput {
  target_type: string;
  target_id: string;
  design_request_id: string;
  deliverable_id?: string | null;
  step_order: number;
  step_name: string;
  approver_role?: string | null;
  approver_name?: string | null;
  status: ApprovalStatus;
  decision_reason?: string | null;
  due_at?: string | null;
}

export interface ApprovalStepUpdateInput {
  status?: ApprovalStatus;
  approver_name?: string | null;
  decision_reason?: string | null;
  due_at?: string | null;
}

export interface ApprovalStepListResponse {
  data: ApprovalStep[];
  meta: { count: number };
  error: null;
}

File: backend/app/schemas/approval_step.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.models.enums import ApprovalStatus
from app.schemas.common import TimestampedSchema


class ApprovalStepBase(BaseModel):
    target_type: str
    target_id: UUID
    design_request_id: UUID
    deliverable_id: UUID | None = None
    step_order: int
    step_name: str
    approver_role: str | None = None
    approver_name: str | None = None
    status: ApprovalStatus
    decision_reason: str | None = None
    due_at: datetime | None = None


class ApprovalStepCreate(ApprovalStepBase):
    pass


class ApprovalStepUpdate(BaseModel):
    status: ApprovalStatus | None = None
    approver_name: str | None = None
    decision_reason: str | None = None
    due_at: datetime | None = None


class ApprovalStepRead(ApprovalStepBase, TimestampedSchema):
    id: UUID
    decided_at: datetime | None = None


class ApprovalStepListResponse(BaseModel):
    data: list[ApprovalStepRead]
    meta: dict[str, int]
    error: None = None
